File: ruby/locomotive-engineer/locomotive_engineer.rb
class LocomotiveEngineer
  def self.generate_list_of_wagons(*wagon_ids)
    wagon_ids
  end

  def self.fix_list_of_wagons(each_wagons_id, missing_wagons)
    return each_wagons_id[2], *missing_wagons, *each_wagons_id[3..], *each_wagons_id[..1]
  end

  def self.add_missing_stops(route, **stops)
    route.merge({ :stops => stops.values })
  end

  def self.extend_route_information(route, more_route_information)
    route.merge(more_route_information)
  end
end

File: python/locomotive-engineer/locomotive_engineer.py
"""Functions which helps the locomotive engineer to keep track of the train."""


def get_list_of_wagons(*wagon_ids):
    """Return a list of wagons.

    :param wagon_ids: tuple - arbitrary number of wagons.
    :return: list - list of wagons.
    """
    return list(wagon_ids)


def fix_list_of_wagons(each_wagons_id, missing_wagons):
    """Fix the list of wagons.

    :param each_wagons_id: list - the list of wagons.
    :param missing_wagons: list - the list of missing wagons.
    :return: list - list of wagons.
    """
    first, second, engine, *remainder = each_wagons_id
    return [engine, *missing_wagons, *remainder, first, second]


def add_missing_stops(route, **stops):
    """Add missing stops to route dict.

    :param route: dict - the dict of routing information.
    :param stops: dict - arbitrary number of stops.
    :return: dict - updated route dictionary.
    """
    return {**route, 'stops': list(stops.values())}


def extend_route_information(route, more_route_information):
    """Extend route information with more_route_information.

    :param route: dict - the route information.
    :param more_route_information: dict -  extra route information.
    :return: dict - extended route information.
    """
    return {**route, **more_route_information}


def fix_wagon_depot(wagons_rows):
    """Fix the list of rows of wagons.

    :param wagons_rows: list[list[tuple]] - the list of rows of wagons.
    :return: list[list[tuple]] - list of rows of wagons.
    """
    [[a1, a2, a3], [b1, b2, b3], [c1, c2, c3]] = wagons_rows
    return [[a1, b1, c1], [a2, b2, c2], [a3, b3, c3]]
